firestore-tos-extension/functions/src/index.ts

```diff
@@ -0,0 +1,145 @@
+import * as functions from "firebase-functions";
+import * as admin from "firebase-admin";
+import * as log from "./logs";
+import config from "./config";
+import { user } from "firebase-functions/v1/auth";
+
+if (admin.apps.length === 0) {
+  admin.initializeApp({ projectId: "demo-test" });
+}
+
+const auth = admin.auth();
+const db = admin.firestore();
+
+export const acceptTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      return;
+    }
+
+    /** check if tosId has been provided  */
+    if (!data.tosId) {
+      console.warn("No tosId provided");
+      return;
+    }
+
+    /** check tos documents */
+    const tosDoc = await db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos")
+      .doc(data.tosId)
+      .get();
+
+    /** Return if no agreement exists  */
+    if (!tosDoc || !tosDoc.exists) {
+      console.warn("tosDoc does not exist");
+      return;
+    }
+
+    /** Find current acknowledgements */
+    const acknowledgements = await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(`${process.env.EXT_INSTANCE_ID}`)
+      .get()
+      .then(($) => $.data() || {});
+
+    acknowledgements[data.tosId] = {
+      ...tosDoc.data(),
+      acceptanceDate: admin.firestore.FieldValue.serverTimestamp(),
+    };
+
+    /** Add Acceptance cliam to user document list */
+    await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(`${process.env.EXT_INSTANCE_ID}`)
+      .set({ ...(acknowledgements || {}) });
+
+    /** Set claims on user and return */
+    const claims = {};
+    claims[process.env.EXT_INSTANCE_ID] = acknowledgements;
+    return auth.setCustomUserClaims(context.auth.uid, { ...claims });
```

Comment:
This will remove all claims on this user across all usage. We should merge claims from the existing user object, if possible.

Reply:
Agreed. Will update and add test cases for modifying existing claims.

Reply:
Update added with test cases for appending claims.